File: models/python/models/entities/product.py
"""
Product Entity - IKEA Product Data Model

Represents a product in the IKEA warehouse with location, stock, and pricing information.
Designed for offline-first sync with Couchbase Lite and deserialization from JSON.
"""

from typing import Literal, Optional
from datetime import datetime
from pydantic import BaseModel, Field, field_validator, ConfigDict


class Dimensions(BaseModel):
    """Product dimensions information"""

    length: float = Field(..., description="Length of the product")
    width: float = Field(..., description="Width of the product")
    height: float = Field(..., description="Height of the product")
    unit: str = Field(default="cm", description="Unit of measurement (e.g., 'cm', 'in')")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {"length": 77, "width": 148, "height": 77, "unit": "cm"}
        }
    )


class ProductLocation(BaseModel):
    """Warehouse location information for a product"""

    aisle: int = Field(..., description="Aisle number in warehouse")
    bay: int = Field(..., description="Bay number within aisle")
    section: str = Field(..., description="Section identifier (e.g., 'LIVING_ROOM')")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {"aisle": 4, "bay": 1, "section": "LIVING_ROOM"}
        }
    )


class Product(BaseModel):
    """IKEA Product entity with offline-first support"""

    # Core identity and metadata
    id: str = Field(..., description='Document ID in format "prod_0001"')

    # Basic product information
    article_number: str = Field(
        ..., description='IKEA article number (e.g., "110.969.53")'
    )
    name: str = Field(..., description='Product name (e.g., "STUVA Storage unit")')
    description: str = Field(..., description="Product description")
    category: str = Field(..., description="Product category (e.g., 'Living Room')")
    product_type: str = Field(
        ..., description="Product type (e.g., 'Storage unit', 'Cushion cover')"
    )

    # Pricing and inventory
    price: float = Field(..., gt=0, description="Product price")
    currency: str = Field(..., description='Currency code (e.g., "SEK", "USD")')
    stock: int = Field(..., ge=0, description="Current stock quantity")
    in_stock: bool = Field(..., description="Whether the product is in stock")

    # Physical characteristics
    dimensions: Dimensions = Field(..., description="Product dimensions")
    weight: float = Field(..., ge=0, description="Product weight in kg")
    colors_available: list[str] = Field(
        default_factory=list, description="Available colors for this product"
    )
    assembly_required: bool = Field(
        ..., description="Whether assembly is required"
    )

    # Warehouse location
    store_location: ProductLocation = Field(
        ..., description="Warehouse location information"
    )

    @field_validator("currency")
    @classmethod
    def validate_currency(cls, v):
        """Validate currency code is uppercase 3-letter code"""
        if not v.isupper() or len(v) != 3:
            raise ValueError("Currency must be a 3-letter uppercase code (e.g., SEK, USD)")
        return v

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "id": "prod_0001",
                "article_number": "110.969.53",
                "name": "STUVA Storage unit",
                "description": "Practical and stylish storage unit.",
                "category": "Living Room",
                "product_type": "Storage unit",
                "price": 3430,
                "currency": "SEK",
                "stock": 23,
                "in_stock": True,
                "dimensions": {
                    "length": 77,
                    "width": 148,
                    "height": 77,
                    "unit": "cm",
                },
                "weight": 2.0,
                "colors_available": ["Beige", "Brown", "Grey"],
                "assembly_required": True,
                "store_location": {
                    "aisle": 4,
                    "bay": 1,
                    "section": "LIVING_ROOM",
                },
            }
        },
    )


def create_product_id(article_number: str) -> str:
    """Creates a Product document ID from an article number"""
    return f"product:{article_number}"


def get_article_number_from_id(product_id: str) -> str:
    """Extracts article number from a Product document ID"""
    return product_id.replace("product:", "")

File: models/typescript/entities/product.ts
/**
 * Product Entity - IKEA Product Data Model
 *
 * Represents a product in the IKEA warehouse with location, stock, and pricing information.
 * Designed for offline-first sync with Couchbase Lite.
 */

export interface ProductStock {
  /** Current stock quantity */
  quantity: number;
  /** Stock location identifier (e.g., "LIVING_ROOM", "BEDROOM") */
  location: string;
}

export interface ProductDimensions {
  /** Depth in specified unit */
  depth: number;
  /** Width in specified unit */
  width: number;
  /** Height in specified unit */
  height: number;
  /** Unit of measurement (e.g., "cm", "in") */
  unit: string;
}

export interface Product {
  /** Document ID in format "product:702.758.14" */
  _id: string;

  /** Couchbase revision for conflict resolution */
  _rev?: string;

  /** Document type discriminator */
  type: "product";

  /** IKEA article number (e.g., "702.758.14") */
  articleNumber: string;

  /** Product name (e.g., "KALLAX") */
  name: string;

  /** Product description (e.g., "Shelf unit") */
  description: string;

  /** Product category */
  category: string;

  /** Product type/subcategory */
  productType: string;

  /** Product price */
  price: number;

  /** Currency code (e.g., "SEK", "USD") */
  currency: string;

  /** Stock information */
  stock: ProductStock;

  /** Product dimensions */
  dimensions: ProductDimensions;

  /** Product weight in kg */
  weight: number;

  /** Product tags/attributes */
  tags: string[];

  /** Whether assembly is required */
  assemblyRequired: boolean;

  /** Whether product is in stock */
  inStock: boolean;

  /** ISO 8601 timestamp of last update */
  lastUpdated: string;

  /** ISO 8601 timestamp of last sync with backend */
  _syncedAt?: string;

  /** Flag indicating pending changes to sync */
  _pendingSync?: boolean;
}

/**
 * Type guard to check if an object is a Product
 */
export function isProduct(obj: any): obj is Product {
  return (
    obj &&
    typeof obj._id === 'string' &&
    obj.type === 'product' &&
    typeof obj.articleNumber === 'string' &&
    typeof obj.name === 'string' &&
    typeof obj.description === 'string' &&
    typeof obj.category === 'string' &&
    typeof obj.productType === 'string' &&
    typeof obj.price === 'number' &&
    typeof obj.currency === 'string' &&
    obj.stock &&
    typeof obj.stock.quantity === 'number' &&
    typeof obj.stock.location === 'string' &&
    obj.dimensions &&
    typeof obj.dimensions.depth === 'number' &&
    typeof obj.dimensions.width === 'number' &&
    typeof obj.dimensions.height === 'number' &&
    typeof obj.dimensions.unit === 'string' &&
    typeof obj.weight === 'number' &&
    Array.isArray(obj.tags) &&
    typeof obj.assemblyRequired === 'boolean' &&
    typeof obj.inStock === 'boolean' &&
    typeof obj.lastUpdated === 'string'
  );
}

/**
 * Creates a Product document ID from an article number
 */
export function createProductId(articleNumber: string): string {
  return `product:${articleNumber}`;
}

/**
 * Extracts article number from a Product document ID
 */
export function getArticleNumberFromId(productId: string): string {
  return productId.replace('product:', '');
}
